### PR TITLE
Add support for nix flake workflow.

### PR DESCRIPTION
**Pull request by alejandrosame**
## Description

This PR adds a shim flake that loads shell.nix. Now users will be able to use `nix-shell` and `nix develop` to use the `qmk` CLI to configure their boards.

The changes to the shell.nix code were left to the minimum possible:
  * Code formatted with `alejandra`. (Please let me know if this should be amended with any other formatter for the time being).
  * The python overlay got moved to its own separate file under `util/nix` so it could be properly reused by both `flake.nix` and `shell.nix`. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by sigprof** [comment]
One annoying problem with flakes is https://github.com/NixOS/nix/issues/3121 — until that feature gets implemented in Nix, just adding a `flake.nix` would result in the whole `qmk_firmware` repository getting copied to the Nix store (and even moving `flake.nix` to a subdirectory does not help, because the Nix code will still detect a Git working copy and copy all tracked files into the Nix store).

Also see the previous attempt to convert `shell.nix` to a flake: #17774.

(Note that I did not yet had the time to review this PR properly, just checked that the mentioned Nix issue is still unfixed.)

File: shell.nix [hunk 1]
```diff
@@ -1,88 +1,72 @@
...
+      pythonEnv = poetry2nix.mkPoetryEnv {
+        projectDir = ./util/nix;
+        overrides = poetry2nix.overrides.withDefaults (self: super: {
+          pillow = super.pillow.overridePythonAttrs (old: {
+            # Use preConfigure from nixpkgs to fix library detection issues and
+            # impurities which can break the build process; this also requires
+            # adding propagatedBuildInputs and buildInputs from the same source.
+            propagatedBuildInputs = (old.buildInputs or []) ++ pkgs.python3.pkgs.pillow.propagatedBuildInputs;
+            buildInputs = (old.buildInputs or []) ++ pkgs.python3.pkgs.pillow.buildInputs;
+            preConfigure = (builtins.toString (old.preConfigure or "")) + pkgs.python3.pkgs.pillow.preConfigure;
```
Comment (sigprof) [inline shell.nix:41]: Interesting; I did not need to use `builtins.toString` here — why did you add it?

File: util/nix/sources.nix [hunk 2]
```diff
@@ -1,118 +1,156 @@
 # This file has been generated by Niv.
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply util/nix/sources.nix:1]: True. I'll make the formatter ignore autogenerated files.

File: shell.nix (hunk 1 above)
Comment (alejandrosame) [reply shell.nix:41]: Now that you mention it, I'll test again whether it's needed at this current state.

In my first attempts at making a flake shim, that line was complaining about not receiving any string (although the value of `old.preconfigure` **is** a string). I'll check at this "working state" whether it keeps complaining about it.

**Comment by alejandrosame** [comment]
> One annoying problem with flakes is [NixOS/nix#3121](https://github.com/NixOS/nix/issues/3121) — until that feature gets implemented in Nix, just adding a `flake.nix` would result in the whole `qmk_firmware` repository getting copied to the Nix store (and even moving `flake.nix` to a subdirectory does not help, because the Nix code will still detect a Git working copy and copy all tracked files into the Nix store).
> 
> Also see the previous attempt to convert `shell.nix` to a flake: #17774.
> 
> (Note that I did not yet had the time to review this PR properly, just checked that the mentioned Nix issue is still unfixed.)

Ooops! I do remember reading about that PR at some point. Then, weeks later, I  decided to learn more about nix language, nixpkgs and flakes by working on a software that I use. That's why I came back to do this exercise.

As a nix newbie, all this feedback is valuable to me and I'll implement it. However, I don't want to take more time than necessary from the maintainers interested on nix support, so feel free to close this PR.

Some extra background info: I thought that a flake-shim could be eventually be merged instead of a shell-shim, as I was aware of https://github.com/NixOS/nix/issues/3121. However, my end goal was to make the flake-shim work first, to provide the workflow for those that want it, and then continue with an actual `flake.nix` and shim `shell.nix`, irregardless of the store issue, on my own fork.

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
+  description = "QMK firmware";
+
+  inputs.nixpkgs.url = "github:NixOS/nixpkgs/c0e881852006b132236cbf0301bd1939bb50867e";
```
[the remark this answers is not in the document]
Comment (alejandrosame) [reply flake.nix:4]: I saw some flake flying around that consumes the sources.json from niv to declare the inputs directly. I'll find that example and implement the pattern for this use case.

[the remark this answers is not in the document]
Comment (alejandrosame) [reply flake.nix]: I refactored the code following the general idea of https://github.com/crazazy/niv-flakes. Now, `nix develop` should not even generate a `flake.lock`, and dependencies are loaded by importing the `sources.nix` file as in the previous iteration.

File: util/nix/sources.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (alejandrosame) [reply util/nix/sources.nix]: I reverted the formatting of this file.

File: shell.nix (hunk 1 above)
Comment (alejandrosame) [reply shell.nix]: I took away the call to `builtins.toString` as I didn't find it failing anymore.

File: util/nix/dirty-flake.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
+  description = "QMK firmware dirty flake expression";
+
+  # Load inputs from Niv and make them compatible with outputs args usage.
+  inputs = let
+    packageSources = import ./sources.nix {};
+  in {
+    flake-utils = import packageSources.flake-utils.outPath;
+    nixpkgs = packageSources.nixpkgs.outPath;
+    poetry2nix-src = packageSources.poetry2nix;
```
Comment (alejandrosame) [inline util/nix/dirty-flake.nix:10]: I intended to make this section more generic, but didn't succeed. This manual fiddling made the outputs function get the correct attribute set.
